src/buckets.ts

```diff
@@ -0,0 +1,253 @@
+                path: `/ipfs/${cid.toString()}`,
+                cid: cid,
+                root: cid,
+                remainder: '',
+              },
+              root: event.root?.path ?? '',
+            }
+            resolve(res)
+          } else if (opts?.progress) {
+            opts.progress(event?.bytes)
+          }
+        } else {
+          reject(new Error('Invalid reply'))
+        }
+      })
+      client.onEnd((code) => {
+        if (code === grpc.Code.OK) {
+          resolve()
+        } else {
+          reject(new Error(code.toString()))
+        }
+      })
+      if (source) {
+        const head = new pb.PushPathRequest.Header()
+        head.setPath(source.path || path)
+        head.setKey(key)
+        const req = new pb.PushPathRequest()
+        req.setHeader(head)
+        client.start(this.context.withContext(ctx).toJSON())
+        client.send(req)
+
+        if (source.content) {
+          for await (const chunk of source.content) {
+            const part = new pb.PushPathRequest()
+            part.setChunk(chunk as Buffer)
+            client.send(part)
+          }
+          client.finishSend()
+        }
+      }
+    })
+  }
+
+  /**
+   * Pulls the bucket path, returning the bytes of the given file.
+   * @param key Unique (IPNS compatible) identifier key for a bucket.
+   * @param path A file/object (sub)-path within a bucket.
+   * @param ctx Context object containing web-gRPC headers and settings.
+   * @param opts Options to control response stream. Currently only supports a progress function.
+   */
+  pullPath(
+    key: string,
+    path: string,
+    ctx?: Context,
+    opts?: { progress?: (num?: number) => void },
+  ): AsyncIterableIterator<Uint8Array> {
+    const metadata = this.context.withContext(ctx).toJSON()
+    const chan = new Channel<Uint8Array>()
+    const request = new pb.PullPathRequest()
+    request.setKey(key)
+    request.setPath(path)
+    let written = 0
+    const response = grpc.invoke(API.PullPath, {
+      host: this.serviceHost,
+      transport: this.rpcOptions.transport,
+      debug: this.rpcOptions.debug,
+      request,
+      metadata,
+      onMessage: async (res: pb.PullPathReply) => {
+        const chunk = res.getChunk_asU8()
+        await chan.push(chunk)
+        written += chunk.byteLength
+        if (opts?.progress) {
+          opts.progress(written)
+        }
+      },
+      onEnd: async (status: grpc.Code, message: string, _trailers: grpc.Metadata) => {
+        if (status !== grpc.Code.OK) {
+          throw new Error(message)
+        }
+        await chan.push(Buffer.alloc(0), true)
+      },
+    })
+    return chan.wrap(() => response.close())
+  }
+
+  private unary<
+    R extends grpc.ProtobufMessage,
+    T extends grpc.ProtobufMessage,
+    M extends grpc.UnaryMethodDefinition<R, T>
+  >(methodDescriptor: M, req: R, context?: Context): Promise<T> {
+    // @todo: This is not totally ideal, but is cleaner for returning promises.
```

Comment:
Ideally we'd just use the client directly, but that ends up creating a lot of boilerplate code. So I just do what we've done elsewhere here (but with slightly better type inference here).